src/Microsoft.AspNet.Razor/Parser/HtmlMarkupParser.Block.cs

```diff
@@ -284,15 +294,31 @@ private bool CData()
             }
             else
             {
-                var tagName = String.Empty;
-                if (At(HtmlSymbolType.Text))
+                var tagName = string.Empty;
+                HtmlSymbol bangSymbol = null;
+
+                if (At(HtmlSymbolType.Bang))
+                {
+                    bangSymbol = CurrentSymbol;
+
+                    var nextSymbol = Lookahead();
+
+                    if (nextSymbol != null && nextSymbol.Type == HtmlSymbolType.Text)
+                    {
+                        tagName = "!" + nextSymbol.Content;
+                    }
+                }
+                else if (At(HtmlSymbolType.Text))
                 {
                     tagName = CurrentSymbol.Content;
                 }
+
                 var matched = RemoveTag(tags, tagName, tagStart);
 
                 if (tags.Count == 0 &&
-                    String.Equals(tagName, SyntaxConstants.TextTagName, StringComparison.OrdinalIgnoreCase) &&
+                    // Text tags cannot be targeted by TagHelpers and therefore can't be escaped with '!'.
+                    bangSymbol == null &&
+                    string.Equals(tagName, SyntaxConstants.TextTagName, StringComparison.OrdinalIgnoreCase) &&
```

Comment:
why is `bangSymbol == null` needed?  `tagName` includes the `'!'`, causing the `Equals` to fail when `bangSymbol != null`